app/interactors/datapass_webhook/find_or_create_user.rb

```diff
@@ -1,6 +1,6 @@
 class DatapassWebhook::FindOrCreateUser < ApplicationInteractor
   def call
-    context.user = User.find_or_initialize_by(email: user_attributes['email'])
+    context.user = User.find_or_initialize_by(email: user_attributes['email'].downcase)
```

Comment:
J'introduirais un `find_or_build_from_email` (un truc du genre) pour mettre ce downcase dans la méthode du modèle, ça évite les futurs oubli.

Reply:
bof ça empêchera pas un futur dev de ne pas utiliser cette future méthode alors que là ça va péter avec le `before_save` donc pas besoin d'introduire une méthode

Reply:
> bof ça empêchera pas un futur dev de ne pas utiliser cette future méthode

Si c'est dans la couche modèle je le verrai en faisant la review, si c'est dans un endroit random clairement je le verrai pas.

> alors que là ça va péter avec le before_save donc pas besoin d'introduire une méthode

Raison de plus pour harden

Clairement c'est faible comme implémentation, faut changer d'approche.

Reply:
ça n'apporte pas grand chose de mettre la méthode dans le modèle, ce qu'on veut c'est passer en `downcase` tout ce qui rentre. 

Si on passe en minuscule juste pour la recherche ça veut dire qu'on peut entrer en base "TEST@example.com" puis "test@example.com" va lever une exception...

Pour moi faut laisser le `before_save` sinon la recherche faudrait coder une méthode `find_or_initialize { |u| u.email.downcase == email.downcase }` qui n'existe pas aujourd'hui

Reply:
> ça n'apporte pas grand chose de mettre la méthode dans le modèle, ce qu'on veut c'est passer en downcase tout ce qui rentre.

L'argument que tu avances est exactement celui que j'avance pour le mettre dans le modèle. Un exemple tout simple: si on fait un find ailleurs, faut penser à faire un downcase ?

Btw, ce que je propose est l'approche adoptée par la plupart (voir toute ?) des gems d'authentification (devise par exemple: https://www.rubydoc.info/github/plataformatec/devise/Devise/Models/Authenticatable/ClassMethods#find_for_authentication-instance_method sous le capot ça downcase tout), mais à aucun moment quand tu manipules ta stack d'auth tu fais un downcase de tes params.

Là une implem triviale:

```ruby
  def find_for_authentication(email)
    where('email ilike (?)', email).limit(1).first
  end

  def find_or_initialize_for_authentication(email)
    find_for_authentication(email) || new(email: email)
  end
```

Autre argument contre le callback -> https://medium.com/planet-arkency/the-biggest-rails-code-smell-you-should-avoid-to-keep-your-app-healthy-a61fd75ab2d3 (et un thread qui résume bien mon/mes point(s) de vue là dessus: https://www.reddit.com/r/ruby/comments/cqci3i/callbacks_good_bad_or_ugly/ )

Reply:
Ok si on catch tout ce qui entre/sort autour des emails oui ça va le faire en effet, et avec la contrainte d'unicité on garanti la stabilité de la base. 

Mais par contre faudra être attentif aux prochaines PR qui vont jouer avec les emails car ça pourrait introduire des fonctionnalités qui lèveront des exceptions (du coup on branche Sentry ?)

Et personnellement je ne garanti pas d'être assez attentif pour ne pas les laisser passer, donc Sentry++. C'est en partie pour ça que mon approche était différente je pense.

Je serai ravis de se lancer dans un débat autour des callbacks, j'y suis plutôt favorable en C# on mettait des event partout, c'était la clefs pour découpler, isoler, tester et surtout faire interagir différents éléments de gameplay. Un autre monde d'autres mœurs ? 🤔

Reply:
Dans les derniers commits j'ai introduits les nouvelles méthodes pour chercher un `User` et les ait mises partout où il y avait des `User.find_by_email`

Reply:
> Mais par contre faudra être attentif aux prochaines PR qui vont jouer avec les emails car ça pourrait introduire des fonctionnalités qui lèveront des exceptions (du coup on branche Sentry ?)

Sentry est déjà branché, je solve les erreurs qui poppent au fur et à mesure (i.e. quasiment aucune, vu qu'il n'y a pas de contrainte 🤷)

> Je serai ravis de me lancer dans un débat autour des callbacks, j'y suis plutôt favorable ; en C# on mettait des event partout, c'était la clefs pour découpler, isoler, tester et surtout faire interagir différents éléments de gameplay. Un autre monde d'autres mœurs ? 🤔

L'event based c'est pas exactement la même chose que des callbacks sur des couches modèles (quand tu designs ton app en event based first clairement aucun souci), là par design plus tu mets de callbacks, moins c'est isolé, plus c'est difficile à tester/construire.

Reply:
Sinon en l'état c'est OK pour moi.